src/Workspaces/Core/Portable/Workspace/ProjectSystem/ProjectSystemProjectFactory.cs

```diff
@@ -31,10 +30,13 @@ internal sealed class ProjectSystemProjectFactory
     // serialization lock and then allow us to update our own state under that lock.
     private readonly SemaphoreSlim _gate = new SemaphoreSlim(initialCount: 1);
 
+    private ProjectUpdateState _projectUpdateState = new(ImmutableDictionary<string, ImmutableArray<ProjectId>>.Empty, ImmutableDictionary<ProjectId, ProjectReferenceInformation>.Empty, [], []);
```

Comment:
doc threading rules here.  I assume it's locked by the _gate above.  But we should say it.